# Homework 3

This notebook provides some skeleton code to get you started on the homework.  Add in your own code and markdown cells  to answer the homework questions.  You will submit this notebook as an .ipynb file.  

Before starting, select "Runtime->Factory reset runtime" to start with your directories and environment in the base state.

If you want to save changes to the notebook, select "File->Save a copy in Drive" from the top menu in Colab.  This will save the notebook in your Google Drive.

# Clone, Install, Import

In [ ]:
!git clone https://github.com/zlisto/social_media_genAI

import os
os.chdir("social_media_genAI/main")

Cloning into 'social_media_genAI'...
remote: Enumerating objects: 1872, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 1872 (delta 1), reused 6 (delta 1), pack-reused 1863 (from 2)
Receiving objects: 100% (1872/1872), 209.13 MiB | 13.71 MiB/s, done.
Resolving deltas: 100% (178/178), done.
Updating files: 100% (1496/1496), done.


In [ ]:
pip install -q -r requirements.txt

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 21.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 548.3/548.3 kB 26.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 52.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.0/85.0 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 232.6/232.6 kB 14.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 244.3/244.3 kB 15.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 264.7/264.7 kB 15.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 313.6/313.6 kB 18.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
nx-cugraph-cu12 24.12.0 requires networkx>=3.2, but you have networkx 3.0 which is incompatible.
torch 2.5.1+cu124 requires nvidia-cublas-cu12==12.

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import codecs  #this let's us display tweets properly (emojis, etc.)
import textwrap as tr
from tqdm import tqdm  #progress bar for for loops
from IPython.display import HTML
import json
import textwrap
import plotly.express as px
from sklearn.decomposition import PCA

from scripts.genai import GenAI

#this option makes it so tweets display nicely in a dataframe
pd.set_option("display.max_colwidth", None)

#this code sets the font sizes for plots
plt.rcParams.update({'axes.labelsize': 18,
                     'xtick.labelsize': 14,
                     'ytick.labelsize': 14,
                     'figure.figsize':(8,6),
                     'axes.grid':True})



## Connect to ChatGPT

In [ ]:
OPENAI_API_KEY = ""

In [ ]:
# Initialize GenAI
jarvis = GenAI(OPENAI_API_KEY)

text = "Who are you?"
response = jarvis.generate_text(text)
print(f"Human:{text}\nAI:{response}")

Human:Who are you?
AI:I’m Jarvis, an AI designed to assist you with a wide range of questions and tasks. Whether you need information, advice, or help with problem-solving, I'm here to help! What can I assist you with today?


# Problem 1 (30 points) Generate Tweets in Style of User

We will generate tweets in the style of a user.



## Problem 1.1 (5 points) Load tweets

Load the tweets into a dataframe and call it `df`. The tweets are in the file `"data/TwExportly/TwExportly_deepseek_ai_tweets_2025_01_30.csv"`. Print out the number of rows in `df`.

## Problem 1.2 (5 points) Keep high engagement tweets

Add a column to `df` called `"engagement"` which is the favorite count divided by the view count.  Create two strings called `tweets_str_high` and `tweets_str_low` which are the text of the tweets whose engagment is in the top 10% and bottom 10%.  The string format should be `"Tweet:text\n, Tweet:text\n..."`.

## Problem 1.3 (10 points) Analyze engagement with AI

Create `instructions` which tells the AI return to you an analysis of how the high engagement tweets differ from low engagement tweets.  Also have the instructions show you specific examples of high and low engagement tweets.   Construct your `prompt` so it contains the high and low engagement tweets and distinguishes between them.  Have the AI return the `analysis` as an HTML file.  Display your `analysis` HTML output with dark mode background using the `display` and `HTML` functions so it is easy to read.

## Problem 1.4 (10 points) Enhance tweets with AI

Take the 5 lowest engagement tweets and enhance them with the AI based on your `analysis` from the previous problem.  Put the `analysis`, along with the high and low engagement tweet samples in `instructions`.  Print out the original tweet and its `engagement` value (so we know you picked the low engagement tweets) and then below it print the enhanced tweet.  Make sure you put some kind of demarcation string (like `"-"*50`) between each group of tweets, put headers for the tweets like `"Old Tweet"` and `"Enhanced Tweet"`, and use the `fill` function so it is easier to read.  Using formatted strings, `sort_values` and a `for` loop might be handy here.

# Problem 2 (30 points) Actor and Critic

We will use the AI as a critic to help us improve our tweets.


## Problem 2.1 (5 points) Generte a simple tweet.

Choose a `topic` and a `style` and use the AI to generate a tweet and call it `tweet_base`. Display it with the `display_tweet` function as was done in class (put a semi-colon after the `display_tweet` function so it wont print out the raw HTML for the tweet).

## Problem 2.2 (10 points) Critic

Write a Python function called `critic` that takes as input the tweet text and returns the tweets quality score between 0 and 10, where higher quality means the tweet will get more engagement, the reason why it gave it that score, and ways to improve it.  Have it return the answer as an HTML table in dark model.  Add a doc string to the function so people can get `help` for it.  Print out the `help(critic)` statement.

## Problem 2.3 (5 points) Enhance the base tweet with critic

Use the `critic` function to enhance your `tweet_base` and call it `tweet_enhanced`.  Display the two tweets with `display_tweet` and display the two analyses wiht the `display` and `HTML` functions.   Display the tweet, and then its analysis so its easy to read.

## Problem 2.4 (10 points)  Keep enhancing

Start with `tweet_base` and keep enhancing it with the `critic` function three times.  Display the tweet with `display_tweet` and show the HTML of the `analysis`.

# Problem 3 (10 points) Customer Service Tweets

We will use the AI to make a Twitter customer service bot.

## Problem 3.1 (2 points) Load complaint tweets

Load the data in `"data/tweet_complaints_AmericanAir.csv"` into a dataframe `df_complaints`.  Print the head of the dataframe.

## Problem 3.2 (3 points) Customer service instructions

Create a string `instructions` that tells the AI to address the customer complaints. Create another string `instructions_style` that has information on the identity and style of the AI (name, background, speaking style, etc.  ).  Please be creative with the AI identity.  Add `instructions_style` to `instructions`. Print out `instructions`.

## Problem 3.3 (5 points) Reply to complaints with AI

Use `instructions` and the AI to generate a tweet to reply to each complaint in `df_complaints`.  You can put the complaint in the `prompt`.  Use `display_tweet` to show the complaint tweet and the AI's response tweet below it.  See the AI's `screen_name` to be `"AmericanAirlines"`  Put some space or new lines between each complaint/response pair so it is easy to read.  

# Problem 4 (30 points) Twitter to Instagram

In this problem we will take content from a Twitter account and convert it into content for Instagram

## Problem 4.1 (5 points) Load tweets

Load the tweets from the file `"data/TwExportly/TwExportly_mrbeast_tweets_2024_02_21.csv"` into a dataframe `df`. Define a variable `screen_name` equal to the screen name of the creator of these tweets. Print `screen_name` and the number of tweets in `df`.

## Problem 4.2 (5 points) Most favorite tweet

Sort `df` by `favorite_count` from highest to lowest and display ONLY the  `favorite_count` and `text` columns of the most favorited tweets.

## Problem 4.3 (5 points) Why so popular?

Ask the AI to explain why it thinks the top tweet was the most favorite.  Display the answer as a nicely formatted HTML in dark mode.

## Problem 4.4 (15 points) Image idea from top tweets

Take the top 10 most favorited tweets and give them to the AI in a `prompt`.  Tell it that these are the most liked tweets of the account and to generate a description of an image this account can post on Instagram that will be liked by its followers.  Then generate an image and an Instagram caption from the description.

Display the image and caption as an Instagram post using `display_IG` as was done in class.